cpp/src/streaming/cudf/parquet.cpp

```diff
@@ -153,70 +259,178 @@ Node read_parquet(
             std::ranges::single_view(filter->stream)
         );
     }
-    // TODO: Handle case where multiple ranks are reading from a single file.
-    int files_per_rank =
-        static_cast<int>(files.size() / size + (rank < (files.size() % size)));
-    int file_offset = rank * (files.size() / size) + std::min(rank, files.size() % size);
-    auto local_files = std::vector(
-        files.begin() + file_offset, files.begin() + file_offset + files_per_rank
-    );
     std::uint64_t sequence_number = 0;
     std::vector<std::vector<ChunkDesc>> chunks_per_producer(num_producers);
-    auto const num_files = local_files.size();
-    // Estimate number of rows per file
-    std::size_t files_per_chunk = 1;
-    if (files.size() > 1) {
-        auto nrows =
-            cudf::io::read_parquet_metadata(cudf::io::source_info(local_files[0]))
-                .num_rows();
-        files_per_chunk =
-            static_cast<std::size_t>(std::max(num_rows_per_chunk / nrows, 1l));
-    }
-    auto to_skip = options.get_skip_rows();
-    auto to_read = options.get_num_rows().value_or(std::numeric_limits<int64_t>::max());
-    for (std::size_t file_offset = 0; file_offset < num_files;
-         file_offset += files_per_chunk)
-    {
-        std::vector<std::string> chunk_files;
-        auto const nchunk_files = std::min(num_files - file_offset, files_per_chunk);
-        std::ranges::copy_n(
-            local_files.begin() + static_cast<std::int64_t>(file_offset),
-            static_cast<std::int64_t>(nchunk_files),
-            std::back_inserter(chunk_files)
+    bool rank_has_assigned_work = false;  // Track if this rank was assigned work
+
+    if (files.size() >= size) {
```

Comment:
I think it would be very helpful to move the two cases into separate helper functions, with docstrings describing the two different approaches.